File: src/test/java/mil/army/futures/asitemplate/controllers/TeamControllerTests.java
package mil.army.futures.asitemplate.controllers;


import mil.army.futures.asitemplate.Team;
import mil.army.futures.asitemplate.services.TeamService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;

import java.util.List;

import static org.hamcrest.Matchers.containsString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(TeamController.class)
class TeamControllerTests {

    @Autowired
    private MockMvc mockMvc;

    @MockBean
    private TeamService teamService;

    @Test
    void shouldSaveANewTeamWhenANewTeamIsCreated() throws Exception {
        when(teamService.createTeam(new Team("first-team-name"))).thenReturn(new Team(1L, "first-team-name"));

        this.mockMvc.perform(post("/teams").contentType(MediaType.TEXT_PLAIN).content("first-team-name"))
                .andExpect(status().isCreated())
                .andExpect(content().string(containsString("first-team-name")));

        verify(teamService).createTeam(new Team("first-team-name"));
    }

    @Test
    void shouldRetrieveAllTeamsWhenGettingTeams() throws Exception {
        when(teamService.getTeams()).thenReturn(List.of(new Team(1L, "first-team-name"), new Team(2L, "second-team-name")));

        this.mockMvc.perform(get("/teams"))
                .andExpect(status().isOk())
                .andExpect(content().string(containsString("first-team-name")))
                .andExpect(content().string(containsString("second-team-name")));

        verify(teamService).getTeams();
    }
}

File: src/test/kotlin/mil/army/futures/asitemplate/controllers/TeamControllerTests.kt
package mil.army.futures.asitemplate.controllers

import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.verify
import mil.army.futures.asitemplate.Team
import mil.army.futures.asitemplate.services.TeamService
import org.hamcrest.CoreMatchers.containsString
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post

@WebMvcTest(TeamController::class)
internal class TeamControllerTests {
    @Autowired
    private lateinit var mockMvc: MockMvc

    @MockkBean
    private lateinit var teamService: TeamService

    @Test
    fun `should save a new team when a new team is created`() {
        every { teamService.createTeam(Team(name = "first-team-name")) } returns Team(id = 1L, name = "first-team-name")

        mockMvc.post("/teams") {
            contentType = MediaType.TEXT_PLAIN
            content = "first-team-name"
        }.andExpect {
            status { isCreated() }
            content { string(containsString("first-team-name")) }
        }

        verify(exactly = 1) {
            teamService.createTeam(Team(name = "first-team-name"))
        }
    }

    @Test
    fun `should retrieve all teams when getting teams`() {
        every { teamService.getTeams() } returns listOf(
            Team(id = 1L, name = "first-team-name"),
            Team(id = 2L, name = "second-team-name")
        )

        mockMvc.get("/teams").andExpect {
            status { isOk() }
            content { string(containsString("first-team-name")) }
            content { string(containsString("second-team-name")) }
        }

        verify(exactly = 1) { teamService.getTeams() }
    }

}
